sound/soc/sof/sof-audio.c

```diff
@@ -625,7 +625,7 @@ int sof_set_up_pipelines(struct snd_sof_dev *sdev, bool verify)
 	int ret;
 
 	/* restore pipeline components */
-	list_for_each_entry_reverse(swidget, &sdev->widget_list, list) {
+	list_for_each_entry(swidget, &sdev->widget_list, list) {
```

Comment:
@plbossart I think this is fine with the latest FW but I am not sure if this would cause issues with the previous FW. I think this one is better to be kept as is.

Reply:
I don't get what the firmware might do differently. This is about about walking through a list that includes all the pipelines and widgets. This does not proceed pipeline by pipeline, and the list might be dependent on how the topology was generated. We have no control at the driver level on any order, so the 'reverse' option doesn't see to bring us any value.

Reply:
@plbossart when parsing topology widgets, we use list_add() to add the widgets that are parsed. So if we traverse the list head first, we'll end up with issues when setting up the pipeline widget with older firmware. Until recently (when we made the change in the FW to not set the pipeline scheduling comp until pipeline_complete), we always checked if the scheduling comp exists when initializing a pipeline widget. And this part will fail if we dont traverse in reverse order.
With newer FW it is not a problem.

Reply:
fixed by adding explicit dependencies for scheduler widgets.